server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -814,7 +814,7 @@ void finishAsFailed(Exception failure) {
                 logger.trace(() -> new ParameterizedMessage("operation failed. action [{}], request [{}]", actionName, request), failure);
                 listener.onFailure(failure);
             } else {
-                assert false : "finishAsFailed called but operation is already finished";
+                assert false : new AssertionError("finishAsFailed called but operation is already finished", failure);
```

Comment:
Perhaps use Assertions.ENABLED?
```
if (Assertions.ENABLED) { 
  throw new AssertionError(msg, failure); 
}
```

Reply:
This could work out, but I anticipate the same pattern being used also when the assertion condition is not simply `false` and therefore I would prefer to keep it like it is. I also prefer the assert style because it is easier to read and takes up less lines of code.

Reply:
Yes, fair point.